python/ray/autoscaler/_private/aws/node_provider.py

```diff
@@ -192,7 +157,54 @@ def set_node_tags(self, node_id, tags):
             except KeyError:
                 self.tag_cache_pending[node_id] = tags
 
-            self.tag_cache_update_event.set()
+        # If on main thread, update tags now.
+        if threading.current_thread() is threading.main_thread():
+            self._update_node_tags()
+        else:  # Else, wait for tag_update_thread to update a batch of tags
+            with self.tag_update_thread_lock:
+                if not self.tag_batching_in_progress.is_set():
+                    self.tag_update_thread.join(
+                    )  # Wait for last batch to finish updating
+                    self.tag_update_thread = threading.Thread(
+                        target=self._batched_tag_update)
+                    self.tag_update_thread.start()  # Start the new batch
+            self.tag_update_thread.join(
+            )  # Wait for update to complete before returning
```

Comment:
nit: comment before code to avoid odd line-breaks

Reply:
Noted.

eh... come to think of it, there could be a situation where the update_thread keeps getting started (lines 169/170) and prevents set_node_update from returning (line 171) when it's supposed to.

I need to think about this more carefully.